### DISCUSSION TITLE
How can I create and drop test database during tests?

### DISCUSSION THREAD

**Discussion by arielpontes** [fastapi/fastapi#8572] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

Here's `conftest.py` where I attempt to define a fixture that takes care of creating a test PostgreSQL database in the beginning of the test session, and dropping it after the test suite is done running:

```python
Base = declarative_base()


@pytest.fixture(scope="session")
def db_engine():
    default_db = (
        "postgresql+asyncpg://postgres:postgres@postgres:5432/postgres"
    )
    test_db = "postgresql+asyncpg://postgres:postgres@postgres:5432/test"
    engine = create_async_engine(default_db)
    if not database_exists(test_db):  # <- Getting error on this line
        create_database(test_db)

    Base.metadata.create_all(bind=engine)
    yield engine


@pytest.fixture(scope="function")
def db(db_engine):
    connection = db_engine.connect()

    # begin a non-ORM transaction
    connection.begin()

    # bind an individual Session to the connection
    Session = sessionmaker(bind=connection)
    db = Session()
    # db = Session(db_engine)

    yield db

    db.rollback()
    connection.close()


@pytest.fixture(scope="function")
def client(db):
    app.dependency_overrides[get_session] = lambda: db

    PREFIX = "/api/v1/my-endpoint"
    with TestClient(PREFIX, app) as c:
        yield c
```

Here's the error I get:

```
E                       sqlalchemy.exc.PendingRollbackError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: Task <Task pending name='anyio.from_thread.BlockingPortal._call_func' coro=<BlockingPortal._call_func() running at /usr/local/lib/python3.9/site-packages/anyio/from_thread.py:187> cb=[TaskGroup._spawn.<locals>.task_done() at /usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py:629]> got Future <Future pending cb=[Protocol._on_waiter_completed()]> attached to a different loop (Background on this error at: https://sqlalche.me/e/14/7s2a)

/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py:601: PendingRollbackError
```

### Description

I'm trying to create and drop test database during tests but I couldn't find any resources helping me do that. I posted to questions on StackOverflow ([this](https://stackoverflow.com/questions/70752806/how-to-use-postgresql-test-database-in-async-fastapi-tests) and [this](https://stackoverflow.com/questions/70798578/pendingrollbackerror-when-accessing-test-database-in-fastapi-async-test)) but got no answer. I believe this is a common use-case, so I figure I'd ask here. If this is possible, I believe an example of how to do it should be added to the documentation.

### Operating System

macOS

### FastAPI Version

0.70.1

### Python Version

3.9.9

**Comment by ugtar** [comment]
I have a small project I've been working on the last few years for just
such a use case. You may find it helpful. It's called pg_temp and it is
designed to create throw away userland postgres databases, mainly for
testing purposes. https://pypi.org/project/pg-temp/

-- 
   Uri

Please consider the environment before printing this message.
http://wwf.panda.org/savepaper/

**Comment by jgould22** [comment]
This is more of an sqlalchemy question than FastAPI, you will probably get more help from them.

That said it looks like you are trying to use that sqlalchmy-utils lib with an async client and it probably doesn't support it.

You init the engine right there so you might as well just open an async connection and check for the db there yourself.

**Comment by rafsaf** [comment]
That is correct, question is more sqlalchemy related but I belive can be also answered here. You are almost there, but need to add something like that, `drop_all` and then `create_all`  ;)

```python
@pytest.fixture(scope="session")
async def test_db_setup_sessionmaker():
    # assert if we use TEST_DB URL for 100%
    # some other asserts...
    # async_engine = create_async_engine(...)

    # always drop and create test db tables between tests session
    async with async_engine.begin() as conn:

        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)
    return async_session


@pytest.fixture
async def session(test_db_setup_sessionmaker) -> AsyncGenerator[AsyncSession, None]:
    async with test_db_setup_sessionmaker() as session:
        yield session

```
Note it's not full working example, It is copied from [conftest.py](https://github.com/rafsaf/minimal-fastapi-postgres-template/blob/main/%7B%7Bcookiecutter.project_name%7D%7D/template_minimal/app/tests/conftest.py) in my FastAPI [template](https://github.com/rafsaf/minimal-fastapi-postgres-template), you can check the "working" async flow if you want over there.

**Comment by ghost** [comment]
I think there might be something here, that is not related to sqlalchemy but rather asyncio.

If you look at the stacktrace more in detail it says:
```
Original exception was: Task <Task pending name='anyio.from_thread.BlockingPortal._call_func' coro=<BlockingPortal._call_func() running at /usr/local/lib/python3.9/site-packages/anyio/from_thread.py:187> cb=[TaskGroup._spawn.<locals>.task_done() at /usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py:629]> got Future <Future pending cb=[Protocol._on_waiter_completed()]> attached to a different loop
```

But that might be duplicate to: https://github.com/tiangolo/fastapi/issues/4100
